pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/strategy/SegmentAssignmentStrategyFactory.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.assignment.segment.strategy;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.helix.HelixManager;
+import org.apache.pinot.common.assignment.InstancePartitions;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.config.table.assignment.SegmentAssignmentConfig;
+import org.apache.pinot.spi.utils.CommonConstants.Segment.AssignmentStrategy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Factory for SegmentAssignmentStrategy
+ */
+public class SegmentAssignmentStrategyFactory {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentAssignmentStrategyFactory.class);
+  private SegmentAssignmentStrategyFactory() {
+  }
+
+  /**
+   * Determine Segment Assignment strategy
+   */
+  public static Map<InstancePartitionsType, SegmentAssignmentStrategy>
+  getSegmentAssignmentStrategy(HelixManager helixManager,
```

Comment:
I feel that this logic becomes too complex if we also need to handle all of `OFFLINE/CONSUMING/COMPLETED`. I think that we should consider keeping this factory to be very thin (given an `AssignmentStrategy` enum, return the correct object).

Instead, we should think about putting the logic for picking the correct assignment in each `OfflineSegmentAssignment / RealtimeSegmentAssignment`.

This will make this logic less complex because we just need to handle `OFFLINE` from `OfflineSegmentAssignment` and `CONSUMING/COMPLETED` from `RealtimeSegmentAssignment`.

Reply:
I think both of them share common code for determining replication/ partition information. This would have duplicate code for generating strategy object in both Offline and RealtimeSegmentAssignment

Reply:
We don't want to pre-compute the strategy for all types. Instead, we can simplify it to only return the strategy for one type/tier: `SegmentAssignmentStrategy getSegmentAssignmentStrategy(HelixManager helixManager, TableConfig tableConfig, InstancePartitionsType instancePartitionsType, InstancePartitions instancePartitions)`

Reply:
In the real time tables, we would want to get assignment strategies for both CONSUMING and COMPLETED segments at one time, so the idea was to keep a map. If we change this to fetch single assignment strategy, then probably, we would want to call same function for 2 different instance partition types, instead we can populate map and use the key to determine strategy.

cc @snleee

Reply:
@GSharayu I think that @Jackie-Jiang's suggestion is simpler. 

```
--> OfflineSegmentAssignment

strategy = getSegmentAssignmentStrategy(OFFLINE)
// deal with oflfine
  
--> RealtimeSegmentAssignment

strategy = getSegmentAssignmentStrategy(COMPLETED)
// deal with completed

...
strategy = getSegmentAssignmentStrategy(CONSUMING)
// deal with consuming

```

Reply:
updated